File: ts/earthbucks-iso-buf/src/iso-buf-error.ts
import { Result, Ok, Err } from "earthbucks-opt-res";
import { Option, Some, None } from "earthbucks-opt-res";

export abstract class IsoBufError extends Error {
  constructor() {
    super();
  }
}

export class GenericError extends IsoBufError {
  constructor(
    public source: Option<IsoBufError>,
    public message: string,
  ) {
    super();
  }

  toString(): string {
    return `isobuf error: ${this.message}`;
  }
}

export class NotEnoughDataError extends IsoBufError {
  constructor(public source: Option<IsoBufError>) {
    super();
  }

  toString(): string {
    return `not enough bytes in the buffer to read`;
  }
}

export class TooMuchDataError extends IsoBufError {
  constructor(public source: Option<IsoBufError>) {
    super();
  }

  toString(): string {
    return `too many bytes in the buffer to read`;
  }
}

export class NonMinimalEncodingError extends IsoBufError {
  constructor(public source: Option<IsoBufError>) {
    super();
  }

  toString(): string {
    return `non-minimal encoding`;
  }
}

export class InsufficientPrecisionError extends IsoBufError {
  constructor(public source: Option<IsoBufError>) {
    super();
  }

  toString(): string {
    return `number too large to retain precision`;
  }
}

export class InvalidHexError extends IsoBufError {
  constructor(public source: Option<IsoBufError>) {
    super();
  }

  toString(): string {
    return `invalid hex`;
  }
}

export class InvalidEncodingError extends IsoBufError {
  constructor(public source: Option<IsoBufError>) {
    super();
  }

  toString(): string {
    return `invalid encoding`;
  }
}

File: rs/earthbucks_iso_buf/src/iso_buf_error.rs
use std::fmt;

#[derive(Debug)]
pub enum IsoBufError {
    GenericError {
        source: Option<Box<IsoBufError>>,
        message: String,
    },
    NotEnoughDataError {
        source: Option<Box<IsoBufError>>,
    },
    TooMuchDataError {
        source: Option<Box<IsoBufError>>,
    },
    NonMinimalEncodingError {
        source: Option<Box<IsoBufError>>,
    },
    InsufficientPrecisionError {
        source: Option<Box<IsoBufError>>,
    },
    InvalidHexError {
        source: Option<Box<IsoBufError>>,
    },
    InvalidEncodingError {
        source: Option<Box<IsoBufError>>,
    },
}

impl fmt::Display for IsoBufError {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self {
            IsoBufError::GenericError { message, .. } => {
                write!(f, "isobuf error: {}", message)
            }
            IsoBufError::NotEnoughDataError { .. } => {
                write!(f, "not enough bytes in the buffer to read")
            }
            IsoBufError::TooMuchDataError { .. } => {
                write!(f, "too many bytes in the buffer to read")
            }
            IsoBufError::NonMinimalEncodingError { .. } => {
                write!(f, "non-minimal encoding")
            }
            IsoBufError::InsufficientPrecisionError { .. } => {
                write!(f, "number too large to retain precision")
            }
            IsoBufError::InvalidHexError { .. } => {
                write!(f, "invalid hex")
            }
            IsoBufError::InvalidEncodingError { .. } => {
                write!(f, "invalid encoding")
            }
        }
    }
}
